File: jarz_pos/api/customer_analytics.py
"""
Customer Analytics API — Jarz POS ERPNext Desk page.

Surfaces the RFM segmentation that already runs nightly
(``jarz_pos.services.rfm_segmentation``).  Segment counts and per-customer RFM
metrics are **current-state** (stored on the Customer DocType); revenue figures
are **date-scoped** through ``tabSales Invoice``.

Data sources:
  tabCustomer       — customer_segment, rfm_recency_days, rfm_frequency_count,
                      rfm_avg_order_value (populated by the nightly RFM job)
  tabSales Invoice  — grand_total, posting_date (docstatus=1, is_return=0)

Read-only.  Requires the JARZ Manager role.
"""

from __future__ import annotations

from datetime import date, timedelta
from typing import Any, Dict, Optional

import frappe
from frappe import _

from jarz_pos.constants import ROLES

# Canonical display order for the 8 RFM segments + fallbacks.
SEGMENT_ORDER = [
    "Champion",
    "Loyal",
    "Potential Loyalist",
    "New Customer",
    "At Risk",
    "Can't Lose Them",
    "Lost",
    "One-Time",
    "Unclassified",
]

# Segments surfaced in the "needs attention" list.
AT_RISK_SEGMENTS = ["At Risk", "Can't Lose Them"]


def _ensure_jarz_manager() -> None:
    roles = set(frappe.get_roles(frappe.session.user))
    if ROLES.JARZ_MANAGER not in roles and ROLES.ADMINISTRATOR not in roles:
        frappe.throw(_("Only JARZ Manager can access customer analytics"), frappe.PermissionError)


def _default_dates() -> tuple[str, str]:
    today = date.today()
    return (today - timedelta(days=29)).isoformat(), today.isoformat()


def _segment_sort_key(segment: str) -> int:
    try:
        return SEGMENT_ORDER.index(segment)
    except ValueError:
        return len(SEGMENT_ORDER)


@frappe.whitelist()
def get_customer_analytics(
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
) -> Dict[str, Any]:
    """Aggregated customer / RFM analytics for the given date range.

    Response shape:
    {
      "period": {"date_from": str, "date_to": str},
      "summary": {
        "total_customers": int, "active_in_period": int, "new_customers": int,
        "returning_customers": int, "repeat_rate": float,
        "champions": int, "at_risk": int, "lost": int,
        "period_revenue": float, "period_orders": int, "avg_order_value": float
      },
      "segment_distribution": [{"segment": str, "count": int}],
      "segment_table": [{"segment", "customers", "active_customers",
                         "orders", "revenue", "avg_recency",
                         "avg_frequency", "avg_aov"}],
      "top_customers": [{"customer", "customer_name", "segment",
                         "recency_days", "orders", "revenue"}],
      "at_risk_customers": [{"customer", "customer_name", "segment",
                             "recency_days", "frequency", "avg_aov", "territory"}],
      "acquisition_trend": [{"date": str, "new_customers": int}]
    }
    """
    _ensure_jarz_manager()

    if not date_from or not date_to:
        date_from, date_to = _default_dates()

    params = {"fd": date_from, "td": date_to}

    # ── Snapshot: customer count + avg RFM metrics per segment ───────────
    snapshot_rows = frappe.db.sql(
        """
        SELECT
            COALESCE(customer_segment, 'Unclassified') AS segment,
            COUNT(*)                     AS customers,
            AVG(rfm_recency_days)        AS avg_recency,
            AVG(rfm_frequency_count)     AS avg_frequency,
            AVG(rfm_avg_order_value)     AS avg_aov
        FROM `tabCustomer`
        WHERE disabled = 0
        GROUP BY segment
        """,
        as_dict=True,
    )
    snapshot_map = {r["segment"]: r for r in snapshot_rows}

    # ── Date-scoped revenue per segment ──────────────────────────────────
    revenue_rows = frappe.db.sql(
        """
        SELECT
            COALESCE(c.customer_segment, 'Unclassified') AS segment,
            COUNT(*)                        AS orders,
            COALESCE(SUM(si.grand_total), 0) AS revenue,
            COUNT(DISTINCT si.customer)     AS active_customers
        FROM `tabSales Invoice` si
        JOIN `tabCustomer` c ON c.name = si.customer
        WHERE si.docstatus = 1
          AND si.is_return = 0
          AND si.posting_date BETWEEN %(fd)s AND %(td)s
        GROUP BY segment
        """,
        params,
        as_dict=True,
    )
    revenue_map = {r["segment"]: r for r in revenue_rows}

    # ── Merge snapshot + revenue into one ordered segment table ──────────
    all_segments = set(snapshot_map) | set(revenue_map)
    segment_table = []
    for seg in sorted(all_segments, key=_segment_sort_key):
        snap = snapshot_map.get(seg, {})
        rev = revenue_map.get(seg, {})
        segment_table.append({
            "segment": seg,
            "customers": int(snap.get("customers") or 0),
            "active_customers": int(rev.get("active_customers") or 0),
            "orders": int(rev.get("orders") or 0),
            "revenue": round(float(rev.get("revenue") or 0), 2),
            "avg_recency": round(float(snap.get("avg_recency") or 0), 1),
            "avg_frequency": round(float(snap.get("avg_frequency") or 0), 1),
            "avg_aov": round(float(snap.get("avg_aov") or 0), 2),
        })

    segment_distribution = [
        {"segment": s["segment"], "count": s["customers"]}
        for s in segment_table if s["customers"] > 0
    ]

    # ── Top customers by revenue in range ────────────────────────────────
    top_customers = frappe.db.sql(
        """
        SELECT
            si.customer,
            c.customer_name,
            COALESCE(c.customer_segment, 'Unclassified') AS segment,
            c.rfm_recency_days              AS recency_days,
            COUNT(*)                        AS orders,
            COALESCE(SUM(si.grand_total), 0) AS revenue
        FROM `tabSales Invoice` si
        JOIN `tabCustomer` c ON c.name = si.customer
        WHERE si.docstatus = 1
          AND si.is_return = 0
          AND si.posting_date BETWEEN %(fd)s AND %(td)s
        GROUP BY si.customer, c.customer_name, c.customer_segment, c.rfm_recency_days
        ORDER BY revenue DESC
        LIMIT 20
        """,
        params,
        as_dict=True,
    )
    for r in top_customers:
        r["revenue"] = round(float(r["revenue"] or 0), 2)
        r["orders"] = int(r["orders"] or 0)
        r["recency_days"] = int(r["recency_days"] or 0)

    # ── Customers needing attention (current snapshot) ───────────────────
    at_risk_customers = frappe.db.sql(
        """
        SELECT
            name           AS customer,
            customer_name,
            customer_segment AS segment,
            rfm_recency_days   AS recency_days,
            rfm_frequency_count AS frequency,
            rfm_avg_order_value AS avg_aov,
            territory
        FROM `tabCustomer`
        WHERE disabled = 0
          AND customer_segment IN %(segs)s
        ORDER BY rfm_avg_order_value DESC
        LIMIT 50
        """,
        {"segs": tuple(AT_RISK_SEGMENTS)},
        as_dict=True,
    )
    for r in at_risk_customers:
        r["recency_days"] = int(r["recency_days"] or 0)
        r["frequency"] = int(r["frequency"] or 0)
        r["avg_aov"] = round(float(r["avg_aov"] or 0), 2)

    # ── New-customer acquisition in range (first-ever order in window) ────
    acquisition_trend = frappe.db.sql(
        """
        SELECT f.first_date AS date, COUNT(*) AS new_customers
        FROM (
            SELECT customer, MIN(posting_date) AS first_date
            FROM `tabSales Invoice`
            WHERE docstatus = 1 AND is_return = 0 AND customer IS NOT NULL
            GROUP BY customer
        ) f
        WHERE f.first_date BETWEEN %(fd)s AND %(td)s
        GROUP BY f.first_date
        ORDER BY f.first_date
        """,
        params,
        as_dict=True,
    )
    for r in acquisition_trend:
        r["date"] = str(r["date"])
        r["new_customers"] = int(r["new_customers"] or 0)

    new_customers = sum(r["new_customers"] for r in acquisition_trend)

    # ── Period totals ────────────────────────────────────────────────────
    totals = frappe.db.sql(
        """
        SELECT
            COUNT(*)                        AS orders,
            COALESCE(SUM(grand_total), 0)    AS revenue,
            COUNT(DISTINCT customer)        AS active
        FROM `tabSales Invoice`
        WHERE docstatus = 1
          AND is_return = 0
          AND posting_date BETWEEN %(fd)s AND %(td)s
        """,
        params,
        as_dict=True,
    )[0]

    period_orders = int(totals["orders"] or 0)
    period_revenue = round(float(totals["revenue"] or 0), 2)
    active_in_period = int(totals["active"] or 0)
    avg_order_value = round(period_revenue / period_orders, 2) if period_orders else 0.0
    returning_customers = max(0, active_in_period - new_customers)
    repeat_rate = round(returning_customers / active_in_period * 100, 1) if active_in_period else 0.0

    def _seg_count(seg: str) -> int:
        return int(snapshot_map.get(seg, {}).get("customers") or 0)

    total_customers = sum(int(r.get("customers") or 0) for r in snapshot_rows)

    summary = {
        "total_customers": total_customers,
        "active_in_period": active_in_period,
        "new_customers": new_customers,
        "returning_customers": returning_customers,
        "repeat_rate": repeat_rate,
        "champions": _seg_count("Champion"),
        "at_risk": _seg_count("At Risk") + _seg_count("Can't Lose Them"),
        "lost": _seg_count("Lost"),
        "period_revenue": period_revenue,
        "period_orders": period_orders,
        "avg_order_value": avg_order_value,
    }

    return {
        "period": {"date_from": date_from, "date_to": date_to},
        "summary": summary,
        "segment_distribution": segment_distribution,
        "segment_table": segment_table,
        "top_customers": top_customers,
        "at_risk_customers": at_risk_customers,
        "acquisition_trend": acquisition_trend,
    }

File: jarz_pos/page/customer_analytics/customer_analytics.js
/* jshint esversion: 9 */
/* globals frappe */

frappe.pages['customer-analytics'].on_page_load = function (wrapper) {
	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Customer Analytics'),
		single_column: true,
	});
	let dash = new CustomerAnalyticsDashboard(page);
	$(wrapper).data('ca_dash', dash);
};

frappe.pages['customer-analytics'].on_page_show = function (wrapper) {
	let dash = $(wrapper).data('ca_dash');
	if (dash) dash.refresh();
};

// ─────────────────────────────────────────────────────────────────────────────

class CustomerAnalyticsDashboard {
	constructor(page) {
		this.page = page;
		this.from_date = frappe.datetime.month_start();
		this.to_date = frappe.datetime.month_end();
		this._setup_filters();
		this._inject_html();
		this.refresh();
	}

	// ── Filters ──────────────────────────────────────────────────────────────

	_setup_filters() {
		let me = this;

		this.page.add_field({
			fieldtype: 'Date',
			fieldname: 'from_date',
			label: __('From'),
			change() {
				me.from_date = me.page.fields_dict.from_date.get_value() || me.from_date;
				me.refresh();
			},
		});
		this.page.fields_dict.from_date.set_value(this.from_date);

		this.page.add_field({
			fieldtype: 'Date',
			fieldname: 'to_date',
			label: __('To'),
			change() {
				me.to_date = me.page.fields_dict.to_date.get_value() || me.to_date;
				me.refresh();
			},
		});
		this.page.fields_dict.to_date.set_value(this.to_date);

		this.page.add_inner_button(__('This Month'), () => {
			this.from_date = frappe.datetime.month_start();
			this.to_date = frappe.datetime.month_end();
			this.page.fields_dict.from_date.set_value(this.from_date);
			this.page.fields_dict.to_date.set_value(this.to_date);
			this.refresh();
		});

		this.page.add_inner_button(__('Last 30 Days'), () => {
			this.to_date = frappe.datetime.nowdate();
			this.from_date = frappe.datetime.add_days(this.to_date, -30);
			this.page.fields_dict.from_date.set_value(this.from_date);
			this.page.fields_dict.to_date.set_value(this.to_date);
			this.refresh();
		});

		this.page.add_inner_button(__('Last 90 Days'), () => {
			this.to_date = frappe.datetime.nowdate();
			this.from_date = frappe.datetime.add_days(this.to_date, -90);
			this.page.fields_dict.from_date.set_value(this.from_date);
			this.page.fields_dict.to_date.set_value(this.to_date);
			this.refresh();
		});
	}

	// ── HTML skeleton ─────────────────────────────────────────────────────────

	_inject_html() {
		this.page.main.html(`
<style>
  .ca { padding: 18px 20px 40px; background: var(--bg-color); }
  .ca-section { margin-bottom: 32px; }
  .ca-title { font-size: 13px; font-weight: 700; color: var(--text-muted);
              text-transform: uppercase; letter-spacing: .6px;
              margin-bottom: 14px; padding-bottom: 8px;
              border-bottom: 1px solid var(--border-color); }
  .ca-note { font-size: 11px; color: var(--text-muted); font-weight: 500;
             text-transform: none; letter-spacing: 0; margin-left: 8px; }

  /* KPI cards */
  .ca-kpis { display: flex; flex-wrap: wrap; gap: 12px; }
  .ca-kpi  { flex: 1; min-width: 140px; background: var(--card-bg);
             border: 1px solid var(--border-color); border-radius: 8px;
             padding: 16px 18px; }
  .ca-kpi .v { font-size: 24px; font-weight: 700; color: var(--text-color); }
  .ca-kpi .l { font-size: 11px; color: var(--text-muted); margin-top: 3px;
               text-transform: uppercase; letter-spacing: .4px; }
  .ca-kpi.pos  .v { color: #27ae60; }
  .ca-kpi.neg  .v { color: #e74c3c; }
  .ca-kpi.info .v { color: #2980b9; }
  .ca-kpi.warn .v { color: #e67e22; }

  /* Grid */
  .ca-grid { display: grid; gap: 18px; grid-template-columns: 1fr 1fr; }
  .ca-grid.one { grid-template-columns: 1fr; }
  .ca-box  { background: var(--card-bg); border: 1px solid var(--border-color);
             border-radius: 8px; padding: 18px; overflow: hidden; }
  .ca-box h6 { font-size: 11px; font-weight: 700; color: var(--text-muted);
               text-transform: uppercase; letter-spacing: .5px; margin-bottom: 14px; }
  .ca-box.span2 { grid-column: 1 / -1; }

  /* Tables */
  .ca-table { width: 100%; border-collapse: collapse; font-size: 13px; }
  .ca-table th { text-align: left; padding: 7px 10px; font-size: 11px; font-weight: 700;
                 text-transform: uppercase; letter-spacing: .4px;
                 color: var(--text-muted); background: var(--subtle-fg);
                 border-bottom: 1px solid var(--border-color); }
  .ca-table td { padding: 7px 10px; border-bottom: 1px solid var(--border-color); }
  .ca-table tr:last-child td { border-bottom: none; }
  .ca-table tr:hover td { background: var(--subtle-fg); }
  .red   { color: #e74c3c; font-weight: 600; }
  .green { color: #27ae60; font-weight: 600; }
  .muted { color: var(--text-muted); }

  .ca-badge { border-radius: 4px; padding: 2px 7px; font-size: 11px; font-weight: 600;
              white-space: nowrap; }

  .ca-loading { text-align: center; color: var(--text-muted); padding: 32px; font-size: 13px; }
</style>

<div class="ca">

  <!-- ① KPIs -->
  <div class="ca-section">
    <div class="ca-title">Key Metrics</div>
    <div class="ca-kpis" id="ca-kpis"><div class="ca-loading">Loading…</div></div>
  </div>

  <!-- ② Segments -->
  <div class="ca-section">
    <div class="ca-title">Customer Segments
      <span class="ca-note">— segment counts are current; revenue is for the selected range</span>
    </div>
    <div class="ca-grid">
      <div class="ca-box">
        <h6>Segment Distribution (Current)</h6>
        <div id="ca-chart-seg-dist"></div>
      </div>
      <div class="ca-box">
        <h6>Revenue by Segment (Selected Range)</h6>
        <div id="ca-chart-seg-rev"></div>
      </div>
      <div class="ca-box span2">
        <h6>Segment Detail</h6>
        <div id="ca-table-seg"></div>
      </div>
    </div>
  </div>

  <!-- ③ Top Customers -->
  <div class="ca-section">
    <div class="ca-title">Top Customers (by Revenue in Range)</div>
    <div class="ca-grid one">
      <div class="ca-box">
        <h6>Highest-Value Customers</h6>
        <div id="ca-table-top"></div>
      </div>
    </div>
  </div>

  <!-- ④ Needs Attention -->
  <div class="ca-section">
    <div class="ca-title">Needs Attention — At Risk &amp; Can't-Lose Customers</div>
    <div class="ca-grid one">
      <div class="ca-box">
        <h6>Win-Back List (Current Snapshot)</h6>
        <div id="ca-table-risk"></div>
      </div>
    </div>
  </div>

  <!-- ⑤ Acquisition -->
  <div class="ca-section">
    <div class="ca-title">New Customer Acquisition</div>
    <div class="ca-grid one">
      <div class="ca-box">
        <h6>First-Time Customers per Day</h6>
        <div id="ca-chart-acq"></div>
      </div>
    </div>
  </div>

</div>`);
	}

	// ── Data loading ─────────────────────────────────────────────────────────

	refresh() {
		$('#ca-kpis').html('<div class="ca-loading">Loading…</div>');

		frappe.call({
			method: 'jarz_pos.api.customer_analytics.get_customer_analytics',
			args: { date_from: this.from_date, date_to: this.to_date },
		}).then(r => {
			if (!r.message) return;
			let d = r.message;
			this._render_kpis(d.summary);
			this._render_segments(d.segment_distribution || [], d.segment_table || []);
			this._render_top(d.top_customers || []);
			this._render_risk(d.at_risk_customers || []);
			this._render_acquisition(d.acquisition_trend || []);
		});
	}

	// ── Renderers ────────────────────────────────────────────────────────────

	_render_kpis(s) {
		let repeat_cls = s.repeat_rate >= 40 ? 'pos' : s.repeat_rate >= 20 ? 'info' : 'warn';
		let risk_cls = s.at_risk > 0 ? 'warn' : 'pos';

		$('#ca-kpis').html(`
			${_kpi('Total Customers',   (s.total_customers || 0).toLocaleString(), '')}
			${_kpi('Active (Range)',    (s.active_in_period || 0).toLocaleString(), 'info')}
			${_kpi('New (Range)',       (s.new_customers || 0).toLocaleString(), 'pos')}
			${_kpi('Repeat Rate',       (s.repeat_rate || 0) + '%', repeat_cls)}
			${_kpi('Champions',         (s.champions || 0).toLocaleString(), 'pos')}
			${_kpi('At Risk / Can\'t Lose', (s.at_risk || 0).toLocaleString(), risk_cls)}
			${_kpi('Lost',              (s.lost || 0).toLocaleString(), 'neg')}
			${_kpi('Avg Order Value',   _egp(s.avg_order_value), 'info')}
		`);
	}

	_render_segments(dist, table) {
		if (dist.length) {
			_chart('ca-chart-seg-dist', {
				type: 'donut',
				height: 280,
				data: {
					labels: dist.map(r => r.segment),
					datasets: [{ values: dist.map(r => r.count) }],
				},
				colors: dist.map(r => _seg_color(r.segment)),
			});
		} else {
			$('#ca-chart-seg-dist').html('<div class="ca-loading">No segmented customers yet.</div>');
		}

		let with_rev = table.filter(r => r.revenue > 0);
		if (with_rev.length) {
			_chart('ca-chart-seg-rev', {
				type: 'bar',
				height: 280,
				data: {
					labels: with_rev.map(r => r.segment),
					datasets: [{ name: 'Revenue', values: with_rev.map(r => r.revenue) }],
				},
				colors: with_rev.map(r => _seg_color(r.segment)),
				tooltipOptions: { formatTooltipY: v => _egp(v) },
			});
		} else {
			$('#ca-chart-seg-rev').html('<div class="ca-loading">No revenue in range.</div>');
		}

		if (!table.length) {
			$('#ca-table-seg').html('<div class="ca-loading">No customer data.</div>');
			return;
		}

		$('#ca-table-seg').html(`
			<table class="ca-table">
				<thead><tr>
					<th>Segment</th>
					<th>Customers</th>
					<th>Active (Range)</th>
					<th>Orders (Range)</th>
					<th>Revenue (Range)</th>
					<th>Avg Recency</th>
					<th>Avg Freq (90d)</th>
					<th>Avg Order Value</th>
				</tr></thead>
				<tbody>
					${table.map(r => `<tr>
						<td><span class="ca-badge" style="background:${_seg_bg(r.segment)};color:${_seg_color(r.segment)}">${r.segment}</span></td>
						<td>${(r.customers || 0).toLocaleString()}</td>
						<td>${(r.active_customers || 0).toLocaleString()}</td>
						<td>${(r.orders || 0).toLocaleString()}</td>
						<td>${r.revenue > 0 ? _egp(r.revenue) : '—'}</td>
						<td class="muted">${r.avg_recency || 0}d</td>
						<td class="muted">${r.avg_frequency || 0}</td>
						<td>${r.avg_aov > 0 ? _egp(r.avg_aov) : '—'}</td>
					</tr>`).join('')}
				</tbody>
			</table>`);
	}

	_render_top(rows) {
		if (!rows.length) {
			$('#ca-table-top').html('<div class="ca-loading">No customer orders in range.</div>');
			return;
		}
		$('#ca-table-top').html(`
			<table class="ca-table">
				<thead><tr>
					<th>#</th>
					<th>Customer</th>
					<th>Segment</th>
					<th>Orders</th>
					<th>Revenue</th>
					<th>Last Order</th>
				</tr></thead>
				<tbody>
					${rows.map((r, i) => `<tr>
						<td class="muted">${i + 1}</td>
						<td><a href="/app/customer/${encodeURIComponent(r.customer)}" target="_blank">${r.customer_name || r.customer}</a></td>
						<td><span class="ca-badge" style="background:${_seg_bg(r.segment)};color:${_seg_color(r.segment)}">${r.segment}</span></td>
						<td>${(r.orders || 0).toLocaleString()}</td>
						<td class="green">${_egp(r.revenue)}</td>
						<td class="muted">${r.recency_days}d ago</td>
					</tr>`).join('')}
				</tbody>
			</table>`);
	}

	_render_risk(rows) {
		if (!rows.length) {
			$('#ca-table-risk').html('<div class="ca-loading">✅ No at-risk customers right now.</div>');
			return;
		}
		$('#ca-table-risk').html(`
			<table class="ca-table">
				<thead><tr>
					<th>Customer</th>
					<th>Segment</th>
					<th>Territory</th>
					<th>Days Since Order</th>
					<th>Freq (90d)</th>
					<th>Avg Order Value</th>
				</tr></thead>
				<tbody>
					${rows.map(r => `<tr>
						<td><a href="/app/customer/${encodeURIComponent(r.customer)}" target="_blank">${r.customer_name || r.customer}</a></td>
						<td><span class="ca-badge" style="background:${_seg_bg(r.segment)};color:${_seg_color(r.segment)}">${r.segment}</span></td>
						<td class="muted">${r.territory || '—'}</td>
						<td class="red">${r.recency_days}d</td>
						<td>${r.frequency}</td>
						<td>${r.avg_aov > 0 ? _egp(r.avg_aov) : '—'}</td>
					</tr>`).join('')}
				</tbody>
			</table>`);
	}

	_render_acquisition(rows) {
		if (!rows.length) {
			$('#ca-chart-acq').html('<div class="ca-loading">No new customers in range.</div>');
			return;
		}
		_chart('ca-chart-acq', {
			type: 'bar',
			height: 280,
			data: {
				labels: rows.map(r => r.date),
				datasets: [{ name: 'New Customers', values: rows.map(r => r.new_customers) }],
			},
			colors: ['#27ae60'],
		});
	}
}

// ─── Helpers ────────────────────────────────────────────────────────────────

function _egp(v) {
	let n = parseFloat(v || 0);
	return 'EGP ' + n.toLocaleString('en', {
		minimumFractionDigits: 0,
		maximumFractionDigits: 0,
	});
}

function _kpi(label, value, cls) {
	return `<div class="ca-kpi ${cls}">
		<div class="v">${value}</div>
		<div class="l">${label}</div>
	</div>`;
}

// Segment colour palette — text/accent colour.
const _SEG_COLORS = {
	'Champion':           '#16a34a',
	'Loyal':              '#2563eb',
	'Potential Loyalist': '#0891b2',
	'New Customer':       '#7c3aed',
	'At Risk':            '#ea580c',
	"Can't Lose Them":    '#dc2626',
	'Lost':               '#6b7280',
	'One-Time':           '#a16207',
	'Unclassified':       '#94a3b8',
};

function _seg_color(segment) {
	return _SEG_COLORS[segment] || '#7B61FF';
}

// Soft background tint for badges (derived from the accent colour at low alpha).
function _seg_bg(segment) {
	let c = _seg_color(segment);
	// Expand #rrggbb to an rgba() with ~14% alpha.
	let r = parseInt(c.slice(1, 3), 16);
	let g = parseInt(c.slice(3, 5), 16);
	let b = parseInt(c.slice(5, 7), 16);
	return `rgba(${r},${g},${b},0.14)`;
}

function _chart(id, opts) {
	let el = document.getElementById(id);
	if (!el) return;
	el.innerHTML = '';
	try {
		new frappe.Chart(el, opts);
	} catch (e) {
		el.innerHTML = `<div class="ca-loading" style="color:#e74c3c">Chart error: ${e.message}</div>`;
	}
}
